src/scene/materials/shader-graph-node.js

```diff
@@ -0,0 +1,614 @@
+        this.genCustomFuncVars();
+    } else {
+        this.graphData.subGraphs = [];
+        this.graphData.connections = [];
+    }
+};
+
+ShaderGraphNode.prototype.constructor = ShaderGraphNode;
+
+Object.assign(ShaderGraphNode.prototype, {
+
+    clone: function () {
+        var clone = new ShaderGraphNode();
+
+        clone.graphData = {};
+
+        clone.graphData.ioPorts = this.graphData.ioPorts.slice(0);
+        clone.graphData.subGraphs = this.graphData.subGraphs.slice(0);
+        clone.graphData.connections = this.graphData.connections.slice(0);
+
+        clone.graphData.customDeclGlsl = this.graphData.customDeclGlsl;
+        clone.graphData.customFuncGlsl = this.graphData.customFuncGlsl;
+
+        return clone;
+    },
+
+    updateShaderGraphUniforms: function (mat) {
+        for (var n = 0; n < this.graphData.ioPorts.length; n++) {
+            var ioPort = this.graphData.ioPorts[n];
+
+            if (ioPort.name.startsWith('IN_') || (ioPort.name.startsWith('CONST_') && ioPort.type === 'sampler2D')) {
+                var chunkId = '_' + this.id;
+
+                switch (ioPort.type) {
+                    case 'sampler2D':
+                        mat.setParameter(ioPort.name + chunkId, ioPort.valueTex);
+                        break;
+                    case 'float':
+                        mat.setParameter(ioPort.name + chunkId, ioPort.valueX);
+                        break;
+                    case 'vec2':
+                        mat.setParameter(ioPort.name + chunkId, [ioPort.valueX, ioPort.valueY]);
+                        break;
+                    case 'vec3':
+                        mat.setParameter(ioPort.name + chunkId, [ioPort.valueX, ioPort.valueY, ioPort.valueZ]);
+                        break;
+                    case 'vec4':
+                        mat.setParameter(ioPort.name + chunkId, [ioPort.valueX, ioPort.valueY, ioPort.valueZ, ioPort.valueW]);
+                        break;
+                    case 'samplerCube':
+                    default:
+                        // error
+                        break;
+                }
+            }
+        }
+    },
+
+    hasValidGraphData: function (graphRootTime) {
+        var i;
+        if (this.graphData.ioPorts && this.graphData.ioPorts.length > 0 && (this.graphData.customFuncGlsl || ( this.graphData.subGraphs && this.graphData.subGraphs.length > 0 && this.graphData.connections))) {
+            for (i = 0; i < this.graphData.ioPorts.length; i++) {
+                if (this.graphData.ioPorts[i] && this.graphData.ioPorts[i].name && this.graphData.ioPorts[i].type) {
+                    if (this.graphData.ioPorts[i].type === 'sampler2D' && !this.graphData.ioPorts[i].name.startsWith('OUT_') ) {
+                        if (!(this.graphData.ioPorts[i].valueTex && this.graphData.ioPorts[i].valueTex instanceof Texture)) {
+                            return false;
+                        }

```

Comment:
you should do something like this:
`var port = this.graphData.ioPorts[i];
and then access it like this:
ports.type ...
`

Reply:
I did exactly that elsewhere - not sure why I missed this loop - now fixed - thanks!